scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,540 @@
+    streams_table_name: String,
+    session: Arc<Session>,
+}
+
+// Number taken from: https://www.scylladb.com/2017/11/17/7-rules-planning-queries-maximum-performance/.
+const DEFAULT_PAGE_SIZE: i32 = 5000;
+
+impl GenerationFetcher {
+    /// Creates new instance of GenerationFetcher.
+    pub fn new(session: &Arc<Session>) -> GenerationFetcher {
+        GenerationFetcher {
+            generations_table_name: "system_distributed.cdc_generation_timestamps".to_string(),
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// In case of a success returns a vector containing all the generations in the database.
+    /// Propagates all errors.
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_all_stream_generations_query());
+        select_consistency(&self.session, &mut query).await?;
+        query.set_page_size(DEFAULT_PAGE_SIZE);
+
+        let mut rows = self
+            .session
+            .query_iter(query, &[])
+            .await?
+            .into_typed::<(chrono::Duration,)>();
+
+        let mut generations = vec![];
+
+        while let Some(next_row) = rows.next().await {
+            let (generation,) = next_row?;
+            generations.push(Generation(generation));
+        }
+
+        Ok(generations)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timestamp_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= ?
+    ORDER BY time DESC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// Given a timestamp of an operation fetch generation that was operating when this operation was performed.
+    /// If no such generation exists, returns None.
+    /// Propagates errors.
+    pub async fn fetch_generation_by_timestamp(
+        &self,
+        time: &chrono::Duration,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_generation_by_timestamp_query());
+        select_consistency(&self.session, &mut query).await?;
+
+        let result = self.session.query(query, (Timestamp(*time),)).await?.rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    /// Given a timeuuid of a ti of an operation fetch generation that was operating when this operation was performed.
+    /// If no such generation exists, returns None.
+    /// Propagates errors.
+    pub async fn fetch_generation_by_timeuuid(
```

Comment:
Do we want to have this function in `GenerationFetcher`'s interface? It doesn't really add any power to this class, it doesn't need any access to `GenerationFetcher`'s internals. Why not leave it to the user of `GenerationFetcher` to extract the timestamps from their timeuuids on their own and just use `fetch_generation_by_timestamp`?

(and if they want to do this often they can easily implement `async fn fetch_generation_by_timeuuid(&GenerationFetcher, &Uuid)`)